toreview/KadActionsBuilder.java

```diff
@@ -0,0 +1,311 @@
+package com.gruppo1.distributednetworkmanager;
+
+import com.dezen.riccardo.networkmanager.StringResource;
+import com.dezen.riccardo.smshandler.SMSPeer;
+
+import java.util.Random;
+
+/**
+ * @author Niccolo' Turcato
+ * Builds KadAction objs
+ * <p>
+ * Based on:
+ * https://github.com/turcato-niccolo/SmsHandler
+ */
+public class KadActionsBuilder implements ActionsBuilder<KadAction, SMSPeer, StringResource, Node<BinarySet>> {
+
+
+    private static int MAX_ACTION_ID;
+    private static int DEFAULT_PARTS = KadAction.MIN_PARTS;
+    private static final String PAYLOAD_IGNORED = "0w0";
+    private static final String ILLEGAL_MAX_ID_MSG = "Can't initialize Builder, the defined max ID isn't compatible with class KadActions";
+    public static final String RESOURCE_SEPARATOR = KadAction.RESOURCE_SEPARATOR;
+
+    public static final KadAction INVALID_ACTION = KadAction.INVALID_KAD_ACTION;
+
+
+    /**
+     * Constructor for this builder, defines a new max limit for IDs, min limit remains the one defined in KadAction
+     *
+     * @param maxActionID max value for action IDs that this builder will accept, must be a valid KadAction ID
+     */
+    public KadActionsBuilder(int maxActionID) {
+        if (maxActionID >= KadAction.MIN_ID && maxActionID <= KadAction.MAX_ID)
+            MAX_ACTION_ID = maxActionID;
+        else
+            throw new IllegalArgumentException(ILLEGAL_MAX_ID_MSG);
+
+    }
+
+    /**
+     * Default constructor, builder will accept all valid IDs, (as indicated in KadAction)
+     */
+    public KadActionsBuilder() {
+        MAX_ACTION_ID = KadAction.MAX_ID;
+    }
+
+    private boolean isValidID(int id) {
+        return id >= KadAction.MIN_ID && id <= MAX_ACTION_ID;
+    }
+
+    /**
+     * @param peer     the peer that will be pinged
+     * @param actionID the unique ID of the action
+     * @return if the given ID is valid: an action built to execute the ping command to the peer, otherwise returns an INVALID_ACTION
+     */
+    public KadAction buildPing(int actionID, SMSPeer peer) {
+        if (isValidID(actionID))
+            return new KadAction(
+                    peer,
+                    KadAction.ActionType.PING,
+                    actionID,
+                    DEFAULT_PARTS, DEFAULT_PARTS,

```

Comment:
Non conosco `KadAction`, ma vedo che capita spesso di passare `DEFAULT_PARTS` _**due**_ volte tra i parametri. Forse dovete rivedere la classe e aggiungere un costruttore che assume i valori di default per questi parametri.

Reply:
Anche, ma il punto è che questo builder andava ad automatizzare la costruzione delle azioni, quindi dove venga fatto ciò mi sembra una sottigliezza.
Il punto è che quando ho scritto questa classe non esisteva questa possibilità (ancora adesso in realtà). Comunque grazie della dritta, vedrò con il mio gruppo se è il caso di implementarlo